File: app/src/main/java/com/jcs/where/adapter/JcsCalendarAdapter.java
package com.jcs.where.adapter;

import android.view.View;
import android.widget.TextView;

import com.chad.library.adapter.base.BaseSectionQuickAdapter;
import com.chad.library.adapter.base.entity.JSectionEntity;
import com.chad.library.adapter.base.viewholder.BaseViewHolder;
import com.jcs.where.R;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.Serializable;
import java.util.List;

import androidx.core.content.ContextCompat;

public class JcsCalendarAdapter extends BaseSectionQuickAdapter<JcsCalendarAdapter.CalendarBean, BaseViewHolder> {
    private boolean mIsEndSelected = false;

    public JcsCalendarAdapter(int sectionHeadResId, int layoutResId, @Nullable List<CalendarBean> data) {
        super(sectionHeadResId, layoutResId, data);
    }


    @Override
    protected void convertHeader(@NotNull BaseViewHolder baseViewHolder, @NotNull CalendarBean calendarBean) {
        baseViewHolder.setText(R.id.yearMonthTv, calendarBean.showYearMonthDate);
    }

    @Override
    protected void convert(@NotNull BaseViewHolder baseViewHolder, CalendarBean calendarBean) {
        View dateView = baseViewHolder.findView(R.id.dateView);
        View leftView = baseViewHolder.findView(R.id.leftView);
        View rightView = baseViewHolder.findView(R.id.rightView);
        // day 为 0 ，说明是空白item
        if (calendarBean.day != 0) {
            baseViewHolder.setText(R.id.dayTv, String.valueOf(calendarBean.day));
            View itemView = baseViewHolder.itemView;
            TextView actionTv = baseViewHolder.findView(R.id.actionTv);
            if (calendarBean.isStartDay) {
                if (mIsEndSelected) {
                    rightView.setBackgroundColor(ContextCompat.getColor(getContext(), R.color.blue_D5EAFF));
                }
                dateView.setBackgroundResource(R.mipmap.start_date_selected);
                actionTv.setText(R.string.enter_stay);
            } else if (calendarBean.isEndDay) {
                leftView.setBackgroundColor(ContextCompat.getColor(getContext(), R.color.blue_D5EAFF));
                dateView.setBackgroundResource(R.mipmap.end_date_selected);
                actionTv.setText(R.string.leave_stay);
            } else if (calendarBean.isSelected) {
                leftView.setBackgroundColor(ContextCompat.getColor(getContext(), R.color.blue_D5EAFF));
                rightView.setBackgroundColor(ContextCompat.getColor(getContext(), R.color.blue_D5EAFF));
                dateView.setBackgroundColor(ContextCompat.getColor(getContext(), R.color.blue_D5EAFF));
                actionTv.setText("");
            } else {
                actionTv.setText("");
                dateView.setBackgroundResource(0);
                leftView.setBackgroundColor(ContextCompat.getColor(getContext(), R.color.white));
                rightView.setBackgroundColor(ContextCompat.getColor(getContext(), R.color.white));
            }
        } else {
            baseViewHolder.setText(R.id.dayTv, "");
            dateView.setBackgroundResource(0);
            dateView.setBackgroundColor(ContextCompat.getColor(getContext(), R.color.white));
            leftView.setBackgroundColor(ContextCompat.getColor(getContext(), R.color.white));
            rightView.setBackgroundColor(ContextCompat.getColor(getContext(), R.color.white));
        }
    }

    public boolean isEndSelected() {
        return mIsEndSelected;
    }

    public void setEndSelected(boolean mIsEndSelected) {
        this.mIsEndSelected = mIsEndSelected;
    }

    public static class CalendarBean extends JSectionEntity implements Serializable {
        long time;
        String showYearMonthDate;
        /**
         * 9月10日
         */
        String showMonthDayDate;
        /**
         * 09-10
         */
        String showMonthDayDateWithSplit;
        String showWeekday;
        int year;
        int month;
        int day;
        boolean isSelected;
        boolean isStartDay;
        boolean isEndDay;
        boolean isToday;
        int itemType;
        boolean isHeader;

        public long getTime() {
            return time;
        }

        public void setTime(long time) {
            this.time = time;
        }

        public String getShowYearMonthDate() {
            return showYearMonthDate;
        }

        public void setShowYearMonthDate(String showYearMonthDate) {
            this.showYearMonthDate = showYearMonthDate;
        }

        public int getYear() {
            return year;
        }

        public void setYear(int year) {
            this.year = year;
        }

        public int getMonth() {
            return month;
        }

        public void setMonth(int month) {
            this.month = month;
        }

        public int getDay() {
            return day;
        }

        public void setDay(int day) {
            this.day = day;
        }

        public boolean isSelected() {
            return isSelected;
        }

        public void setSelected(boolean selected) {
            isSelected = selected;
        }

        public boolean isToday() {
            return isToday;
        }

        public void setToday(boolean today) {
            isToday = today;
        }

        public void setItemType(int itemType) {
            this.itemType = itemType;
        }

        public void setIsHeader(boolean isHeader) {
            this.isHeader = isHeader;
        }

        public boolean isStartDay() {
            return isStartDay;
        }

        public void setStartDay(boolean select) {
            isSelected = select;
            isStartDay = select;
        }

        public boolean isEndDay() {
            return isEndDay;
        }

        public void setEndDay(boolean select) {
            isSelected = select;
            isEndDay = select;
        }

        public String getShowMonthDayDate() {
            return showMonthDayDate;
        }

        public void setShowMonthDayDate(String showMonthDayDate) {
            this.showMonthDayDate = showMonthDayDate;
        }

        public String getShowWeekday() {
            return showWeekday;
        }

        public void setShowWeekday(String showWeekday) {
            this.showWeekday = showWeekday;
        }

        public String getShowMonthDayDateWithSplit() {
            return showMonthDayDateWithSplit;
        }

        public void setShowMonthDayDateWithSplit(String showMonthDayDateWithSplit) {
            this.showMonthDayDateWithSplit = showMonthDayDateWithSplit;
        }

        @Override
        public String toString() {
            return "HotelCalendarBean{" +
                    "time=" + time +
                    ", showYearMonthDate='" + showYearMonthDate + '\'' +
                    ", showMonthDayDate='" + showMonthDayDate + '\'' +
                    ", showWeekday='" + showWeekday + '\'' +
                    ", year=" + year +
                    ", month=" + month +
                    ", day=" + day +
                    ", isSelected=" + isSelected +
                    ", isStartDay=" + isStartDay +
                    ", isEndDay=" + isEndDay +
                    ", isToday=" + isToday +
                    ", itemType=" + itemType +
                    ", isHeader=" + isHeader +
                    '}';
        }

        @Override
        public boolean isHeader() {
            return isHeader;
        }
    }
}

File: app/src/main/java/com/jcs/where/widget/calendar/JcsCalendarAdapter.kt
package com.jcs.where.widget.calendar

import android.graphics.Color
import android.view.View
import android.widget.TextView
import com.blankj.utilcode.util.ColorUtils
import com.chad.library.adapter.base.BaseSectionQuickAdapter
import com.chad.library.adapter.base.entity.SectionEntity
import com.chad.library.adapter.base.viewholder.BaseViewHolder
import com.jcs.where.R
import com.jcs.where.widget.calendar.JcsCalendarAdapter.CalendarBean
import java.io.Serializable

class JcsCalendarAdapter(sectionHeadResId: Int, layoutResId: Int, data: List<CalendarBean>) : BaseSectionQuickAdapter<CalendarBean, BaseViewHolder>(sectionHeadResId, layoutResId, data.toMutableList()) {
    var isEndSelected = false

    override fun convertHeader(helper: BaseViewHolder, item: CalendarBean) {
        helper.setText(R.id.yearMonthTv, item.showYearMonthDate)
    }

    override fun convert(holder: BaseViewHolder, item: CalendarBean) {


        val dateView = holder.getView<View>( R.id.dateView)
        val leftView = holder.getView<View>( R.id.leftView)
        val rightView = holder.getView<View>( R.id.rightView)
        val actionTv = holder.getView<TextView>( R.id.actionTv)


        // day 为 0 ，说明是空白item
        if (item.day != 0) {
            holder.setText(R.id.dayTv, item.day.toString())
            val itemView = holder.itemView
            if (item.isStartDay) {
                if (isEndSelected) {
                    rightView.setBackgroundColor(ColorUtils.getColor(R.color.color_DCE8FF))
                }
                dateView.setBackgroundResource(R.mipmap.start_date_selected)
                actionTv.setText(R.string.calendar_enter_stay)
            } else if (item.isEndDay) {
                leftView.setBackgroundColor(ColorUtils.getColor(R.color.color_DCE8FF))
                dateView.setBackgroundResource(R.mipmap.end_date_selected)
                actionTv.setText(R.string.calendar_leave_stay)
            } else if (item.isSelected) {
                leftView.setBackgroundColor(ColorUtils.getColor(R.color.color_DCE8FF))
                rightView.setBackgroundColor(ColorUtils.getColor(R.color.color_DCE8FF))
                dateView.setBackgroundColor(ColorUtils.getColor(R.color.color_DCE8FF))
                actionTv.text = ""
            } else {
                actionTv.text = ""
                dateView.setBackgroundResource(0)
                leftView.setBackgroundColor(Color.WHITE)
                rightView.setBackgroundColor(Color.WHITE)
            }
        } else {
            holder.setText(R.id.dayTv, "")
            dateView.setBackgroundResource(0)
            dateView.setBackgroundColor(Color.WHITE)
            leftView.setBackgroundColor(Color.WHITE)
            rightView.setBackgroundColor(Color.WHITE)
        }
    }

    class CalendarBean : SectionEntity, Serializable {
        var time: Long = 0

        /**
         * 中文：2021年10月
         * en: Oct 2021
         */
        var showYearMonthDate: String? = null

        /**
         * 中文：9月10日
         * en: Oct 22
         */
        var showMonthDayDate: String? = null

        /**
         * 中文：09-10
         * en: OCT 22
         */
        var showMonthDayDateWithSplit: String? = null

        /**
         * 中文：2021-01-10
         * en: 2021-10-11
         */
        var showYearMonthDayDateWithSplit: String = ""

        /**
         * 中英文：Fri
         */
        var showWeekday: String? = null
        var year = 0
        var month = 0
        var day = 0
        var isSelected = false
        var isStartDay = false
        var isEndDay = false
        var isToday = false
        var mIsHeader = false


        override val isHeader: Boolean
            get() = mIsHeader


//
//        override var itemType = 0
//
//        override var isHeader = false

    }
}